### PR TITLE
[clang][ASTImporter] New fix for default template parameter values.

### PR DESCRIPTION
**Pull request by balazske**
Commit e4440b8 added a change that introduced new crash in an incorrectly handled case. This is fixed here.

### CODE REVIEW

File: clang/unittests/AST/ASTImporterTest.cpp
Comment (balazske) [inline clang/unittests/AST/ASTImporterTest.cpp:9925]: This comment is not accurate, I will remove it.
What exactly should happen:

1. import first `struct X` declaration
3. start import of second `int P`
4. finish import of second `struct X`, this will set `int P` to "inherited"
5. import (visit) function for importing second `int P` returns

**Review by NagyDonat — APPROVED** [review]
This change seems to be correct, but I don't know enough to dig into the details and provide a confident review. However, I as far as I know there are no significantly better reviewers, so I can give you the formal approval.

I think you can merge this commit once you update the comment that you marked as inaccurate.

**Review by NagyDonat — APPROVED** [review]
Feel free to merge this if you're not planning more changes.

The commit seems to be reasonable, I like that your comments explain the logic behind this very technical logic. I don't understand enough to recognize an unhandled corner case (if there was one), but I'm confident that this change is a good step forward.

File: clang/lib/AST/ASTImporter.cpp
Comment (mizvekov) [inline clang/lib/AST/ASTImporter.cpp:5932]:
```suggestion
  // context. This context will be fixed while the actual template
  // declaration is being created.
```

File: clang/lib/AST/ASTImporter.cpp
Comment (mizvekov) [inline clang/lib/AST/ASTImporter.cpp:5979]: I think this avoiding to set a default argument can prevent diagnosing ODR violations, ie different values for the default argument.

In any case, I don't see how it would help to import the default argument, but not set it. The argument would just end up discarded.

The cycle issue we usually solve by deferring importation. I think what could maybe work here is importing all default arguments separately after we are done with the rest of the template.

One thing that was missing from this implementation, and seems to be leading to the wrong source locations issue, is to import the inheritedFrom state.

Ie call `getInheritedFrom` on `D`, import that returned declaration into the `To` context, and set it with `setInherited` on `ToD`.

Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp:5979]: The case when default argument becomes already set happens when it is changed in function `updateTemplateParametersInheritedFrom` that is called when the template is imported. This change can happen during the import of the default argument object. That function calls the `setInheritedFrom` function. I was experimenting with your idea to just call `getInheritedFrom` and import the result and set it (and not use `updateTemplateParametersInheritedFrom`), but I got another crash. I tried to explain this in the added comment.

The "deferring" is done in the current code from `updateTemplateParametersInheritedFrom` that is called at end of the template import. I wanted to ensure that a default value is always set in the template parameter object, because until `updateTemplateParametersInheritedFrom` is called some other import may happen that uses this value. I think this can cause ODR detection problem. If the default argument is not set (but will be set later from `updateTemplateParametersInheritedFrom`) this causes a gap where the default argument is missing. If during this time something is compared to this object it can cause false structural in-equivalence. To avoid this the default argument is imported and set immediately (even if this object is later discarded). Structural equivalence comparison does not compare if the argument is "inherited" or not, but it does compare the argument value.

Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp:5979]: I try to explain the source location problem with this code:
```
template <class P = Def> class X;
template <class P> class X {};
```
At the second `X` (definition) `P` has inherited default argument.
The import may reverse the order (this reversal looks hard to fix):
```
template <class P = Def> class X {};
template <class P> class X;
```
The default argument is now at the definition of `X`. I think this is the good behavior at import and this is how it currently works. The source location of `= Def` becomes inconsistent with the line where it originally was, and this can affect the "end" location of `<class P = Def>` too.

Comment (danix800) [reply clang/lib/AST/ASTImporter.cpp:5979]: I digged a little bit about template importing and have a few questions:

1. Templates are redeclarable, but not like some other nodes (`FunctionDecl`/
    `VarDecl`/`TagDecl` currently), no redecls are imported ahead of the current one.
    if redecls are imported, the ordering problem might go away. If this works,
    we could apply the same mechnism to some other redeclarables too (I tested
    locally and templates can be imported with correct order).

    Or, there might be other considerations about not importing redecls for some
    nodes (like performance? but redecls are usually small, right?), for some client
    (StaticAnalyzer), this mis-ordering is still acceptable, but I'm not quite sure how
    many clients depend on the importer and how they depends on the ordering.

2. For template parameters, current implementation supports importing them directly
    but added it into the top-level TU context, which might not be the good for any
    clients, not to say about ODR violations. I'm not clear what's the consideration
    of not importing some of the nodes without importing context firstly.

For the crash fix it's a LGTM.

Comment (danix800) [reply clang/lib/AST/ASTImporter.cpp:5979]: If template parameter is supposed to be imported only internally (from template itself obviously), we should remove the interface for template parameter, right?

Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp:5979]: In `ASTImporter` there is currently no distinction about what can be imported with the `import` functions. They should work with any object. The caller should ensure that the imported objects are processed correctly (added to something already existing or used to import a bigger object).

Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp:5979]: The order of objects (or the declaration chain) after the import can be different than the original order, probably for any `Decl`. This is because declarations have often recursive or circular references to each other. I think this is true for templates too (and the "templated" declarations in the templates). The redeclarations are imported all the time currently, only if a definition is imported it is mapped to an existing definition if possible. It was an old decision to not merge any (non-definition) declarations at import because it creates only even more difficulties.

Comment (mizvekov) [reply clang/lib/AST/ASTImporter.cpp:5979]: I think it's important preserving the default argument inheritance chain order, by importing it explicitly and disabling the automatic inheritance during import.

The default argument can only be used if it has been declared before use, in source code parsing order.

It seems like it would create difficulties parsing the AST if this order becomes arbitrary.

For example: https://godbolt.org/z/cx9jf5nso

Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp:5979]: The order of redeclarable declarations can change during import, `ASTImporter` worked always this way. In the "order" I mean order got with `getPreviousDecl`. If we go back the decl chain of a template with default argument, the first one should have the non-inherited default argument and all others should have an inherited one. This property is preserved now, even if the "To" AST has different declaration chain than the "From" AST, always the first occurrence of a template will have the non-inherited default argument.

Comment (mizvekov) [reply clang/lib/AST/ASTImporter.cpp:5979]: Yes I understand the redeclaration chain order is not reliable with AST import. It however has no semantic implications.

But making the default argument inheritance chain based on the redeclaration chain order is not helpful, as it will make the AST represent seemingly nonsensical things, like a certain use of a declaration without a default argument succeed in a use which would need it.

The following properties are important to be preserved:
* Declarations which did not have a default argument definition visible should keep so after import.
* The declaration which introduced the default argument definition should keep being so after import.
* Declarations which had the default argument inherited should keep being so after import.

If inherited default arguments always point to the argument with the definition, instead of a previously inherited entry, which I think should already be the case, then this won't create many difficulties here.

Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp]: I made a new fix where the original inheritance order does not change. There are new problems with this solution but probably it is acceptable.
If there is an existing declaration `template <int I = 1> struct X;` and the same code is imported, the declaration is duplicated because it is not a definition (this is how `ASTImporter` generally works). In this case the default value `I = 3` is imported too. The AST is already invalid after this import (default value can be specified only once). With a fix like the previous solution this problem does not exist.
Another problem is with existing code
```
template <int I = 1> struct X {};
```
and imported code
```
template <int I = 1> struct X {};
template <int I> struct X;
```
The second declaration is imported with inherited default argument but the first is not imported, only mapped to the existing one. The "inherited from" value in the second declaration (in the To AST) is not linked to the original one, a new object is created instead that is "invisible" from the AST (only reachable with the `getInheritedFrom` call).

Comment (mizvekov) [reply clang/lib/AST/ASTImporter.cpp]: Thanks. I think this moved in a better direction.

At least this way, the AST is correct for simple reserialization, whereas as before it was incorrect for that as well.

The new problems are related to merging of definitions. We ought to implement a mechanism to merge default argument definitions here the same way we merge function definitions and such.

For your first example: `template <int I = 1> struct X;` Here is one idea: this should be treated as a definition and merged, and any differences would be an ODR violation.

In particular, merging two modules each containing:
```C++
template <int I, int J = 1> struct X;
```
and
```C++
template <int I = 0, int J = 1> struct X;
```
Because the declaration containing a default argument definition differs, this would be an ODR violation.

I'd be happy to leave solving that problem for a future patch.

For your second example, `X` would already be merged, so I am not sure there is a problem there. So we may end up with getInheritedFrom pointing to a definition which was not chosen as the merged definition, but I believe we keep track of the merged definitions in the ASTContext, so we can make sense of the situation.

**Review by mizvekov — APPROVED** [review]
LGTM

**Pull request merged**
